fw/sock_clnt.c

```diff
@@ -474,13 +475,16 @@ tfw_sk_write_xmit(struct sock *sk, struct sk_buff *skb, unsigned int mss_now,
 
 	if (h2_mode && r != -ENOMEM && (flags & SS_F_HTTT2_HPACK_TBL_SZ_ENCODED))
 		tfw_hpack_enc_tbl_write_sz_release(tbl, r);
-	if (unlikely(r) && r != -ENOMEM)
+	if (unlikely(r) && r != -ENOMEM) {
 		/*
 		 * We can not send unencrypted data and can not normally close the
 		 * socket with FIN since we're in progress on sending from the write
 		 * queue.
 		 */
-		ss_close(sk, SS_F_ABORT);
+		SS_CONN_TYPE(sk) |= Conn_Closing;
+		ss_close(sk, SS_F_ABORT_FORCE | __SS_F_FULL_CLOSE);
```

Comment:
It seems I don't understand the main idea of the patch. I understood it as that we split the `ss_do_close()` to 2 parts and kill the socket only when we send all the data on it.

We have `ss_close()` calls also in tls.c and http.c along with sock.c itself - if there is no error (not a case for `ss_linkerror()`) and we have some data on the socket, then `ss_do_close()` is just called on `all other `ss_close()` and if we have some data on socket, then who kills the socket? I see only 3 calls of `ss_finish_closing()`: `ss_do_close()` for sockets with no data, `ss_linkerror()` and `ss_full_close()` which doesn't cover the cases for `ss_close()` calls.